ratings/ratings.py

```diff
@@ -0,0 +1,298 @@
+
+    return ratings
+
+
+def top_n(n=20):
+  rating_tups = compute_ratings()
+  r = compute_ratings()
+  return [(model_num_for(k),v) for v,k in
+          sorted([(v,k) for k,v in r.items()])[-n:][::-1]]
+
+def ingest_dirs(root, dirs):
+    for d in dirs:
+        if os.path.isdir(os.path.join(root,d)):
+            fs = [os.path.join(root, d, f) for f in os.listdir(os.path.join(root, d))]
+            print("Importing from {}".format(d))
+            import_files(fs)
+
+
+def last_timestamp():
+    db = sqlite3.connect("ratings.db")
+    with db:
+        ts = db.execute("select timestamp from games order by timestamp desc limit 1").fetchone()
+    return ts[0] if ts else None
+
+
+def suggest_pairs(top_n=10, per_n=3):
+    """ Find the maximally interesting pairs of players to match up
+    First, sort the ratings by uncertainty.
+    Then, take the ten highest players with the highest uncertainty
+    For each of them, call them `p1`
+    Sort all the models by their distance from p1's rating and take the 20 closest
+    Choose the model with the greatest age difference from among those candidates
+
+    'ratings' is a list of (model_num, rating, uncertainty) tuples
+
+    Returns a list of *model numbers*, not model ids.
+    """
+    db = sqlite3.connect("ratings.db")
+    data = db.execute("select model_winner, model_loser from wins").fetchall()
+    bucket_ids = [id[0] for id in db.execute("select id from models where bucket = ?", (fsdb.models_dir(),)).fetchall()]
+    bucket_ids.sort()
+    data = [d for d in data if d[0] in bucket_ids and d[1] in bucket_ids]
+
+    ratings = [(model_num_for(k), v[0], v[1]) for k,v in compute_ratings(data).items()]
+    ratings.sort()
+    ratings = ratings[70:] # filter off the first 100 models, which improve too fast.
+
+    ratings.sort(key=lambda r: r[2], reverse=True)
+
+    res = []
+    for p1 in ratings[:top_n]:
+        candidate_p2s = list(sorted(ratings, key=lambda p2_tup: abs(p1[1] - p2_tup[1])))[:20]
+        choices = random.sample([c for c in candidate_p2s if c[0] != p1[0]], per_n)
+        print("Pairing {}, sigma {:.2f}".format(p1[0], p1[2]))
+        for p2 in choices:
+            res.append([p1[0], p2[0]])
+            print("   {}, ratings delta {:.2f}".format(p2[0], abs(p1[1]-p2[1])))
+    return res
+
+
+def sync(root, force_all=False):
+    import subprocess
+    last_ts = last_timestamp()
+    if last_ts and not force_all:
+        # Build a list of days from the day before our last timestamp to today
+        num_days = (dt.datetime.utcnow() - dt.datetime.utcfromtimestamp(last_ts) + dt.timedelta(days=1)).days
+        ds = [(dt.datetime.utcnow() - dt.timedelta(days=d)).strftime("%Y-%m-%d") for d in range(num_days+1)] 
+        for d in ds:
+            #TODO: ensure exists os.path.join(root,d)
+            cmd = "gsutil -m rsync -r {0} {1}".format(os.path.join(fsdb.eval_dir(),d), os.path.join(root, d))
+            print(cmd)
+            subprocess.call(cmd.split())
```

Comment:
techincally shlex is preferred but why not just pass the list

cmd = ["gsutil", "-m", "rsync", "-r", X, Y)]
print(" ".join(cmd))
subprocess.call(cmd)

Reply:
sure, done.